### ISSUE TITLE
Re-inventing a wheel

### ISSUE THREAD

**Issue by heisian** [browner12/helpers#2]
Hey man, plz take a look at:
https://github.com/laravel/framework/blob/master/src/Illuminate/Support/ClassLoader.php

It already implements what this package does, except with `ClassLoader` you can load classes from anywhere within your project.

Excerpt:

``` php
    public static function load($class)
    {
        $class = static::normalizeClass($class);
        foreach (static::$directories as $directory) {
            if (file_exists($path = $directory.DIRECTORY_SEPARATOR.$class)) {
                require_once $path;
                return true;
            }
        }
        return false;
    }
```

**Comment by browner12** [comment]
interesting. remember we're not adding classes here, just functions. but I assume it works the same either way. how would you integrate this into a Service Provider, or define the helpers you want `required`?

**Comment by heisian** [comment]
Since the `ClassLoader` is already integrated in the framework and is bootstrapped when the app loads, you only need to create a file somewhere, for example in:
`app/SomeFolder/Another/Subfolder`

and give it a namespace of `namespace App\SomeFolder\Another\Subfolder;`

and the class contained therein can be used anywhere in Laravel 5, automagically.  The `ClassLoader` file will pick it up as long as the directory structure and namespace matches the following PSR-4 spec:
http://www.php-fig.org/psr/psr-4/

You don't even need to run composer dump-autoload 💃

**Comment by heisian** [comment]
Here's an example:
![image](https://cloud.githubusercontent.com/assets/4141828/14863547/f5f59d36-0c6a-11e6-9218-78e375dd3754.png)
TokBox is my custom class for using [OpenTok](https://tokbox.com/developer/)'s WebRTC server SDK.

Now, if I go to my CLI and do `php artisan tinker`, I can directly invoke this class:
![image](https://cloud.githubusercontent.com/assets/4141828/14863609/44badef4-0c6b-11e6-9a4d-f123d3639d66.png)

If `generateSession()` were static, I could call `\App\Core\TokBox\TokBox::generateSession()`.

Using the class within your app in a Controller you could do a `use \App\Core\TokBox\TokBox;` at the top of the file, and then within one of your methods you could simply call
`(new TokBox)->generateSession()`
 or
`TokBox::generateSession()`
if the method was static.

Honestly, you had the right idea here.  Your implementation is very in-line with the Laravel way of doing things.  I just want to let you (& others) know that this type of functionality is already shipped with the framework! :)

**Comment by browner12** [comment]
so if i'm following you correctly, you're talking PSR-4 autoloading, and then calling static methods off that class.

if that's what you're talking about, I agree, that is an option.  However, I want plain vanilla PHP functions, which I don't think is served by this.

You might say that static methods and vanilla functions are essentially the same thing and give the same end result, but there are some performance considerations. Although the results seem a little inconclusive and won't necessarily be the same between PHP versions.

http://stackoverflow.com/questions/1472721/performance-of-static-methods-vs-functions

I also like the fact that vanilla functions are a little more terse than static or object method calls.

`makeButton()`

vs

`\App\Factories\ButtonFactory::makeButton()`

or

``` php
$button = new \App\Factories\ButtonFactory();
$button->makeButton();
```

IMO it can make your code less readable to use the longer ways.

**Comment by heisian** [comment]
Your package is running a **second** iteration of what Laravel **already** does.  Additionally, the performance gains of 0.05 _milliseconds_ demonstrated between static and vanilla function calls are absolutely negligible.  If a programmer was worried about that type of performance, he/she wouldn't be using PHP.

If your intent is to extend the vanilla functions, then I would do just that.  Take a look at how the Laravel framework does it: https://github.com/laravel/framework/blob/master/src/Illuminate/Foundation/helpers.php

Now, in your examples, all it takes is a:

``` php
use \App\Factories\ButtonFactory;
```

statement at the top of your file to shorten all your calls down to:

``` php
ButtonFactory::makeButton();
```

You can even make non-static calls like this:

``` php
(new ButtonFactory)->makeButton();
```

PSR-4 is what makes ALL of Laravel function as an app.  Why circumvent it? You can even write:

``` php
use \App\Factories\ButtonFactory as BF;
```

and shorten your function call to:

``` php
BF::makeButton()
```

Just be glad you're not writing Objective-C...

``` objective-c
-(void)subscriberVideoDisableWarning:(OTSubscriberKit *)subscriber
{
    if ([_delegate respondsToSelector:@selector(videoSessionDidReceiveNetworkWarning:)])
    {
        [_delegate videoSessionDidReceiveNetworkWarning:self];
    }
}
```

To put it simply, I'm trying to advocate a 'best practices' approach here, and one that fits in with the framework we and many others so love.  To the community, I would recommend the built-in PSR-4 autoloader solution over an additional dependency any day.

**Comment by browner12** [comment]
thanks for your input. if you can show me a significant downside to this package or another way to include vanilla php functions, I'd love to listen. otherwise I'm perfectly content with this package and what it provides.

**Comment by heisian** [comment]
The downside to this package (and many others, not just yours) is that additional dependencies provide for additional cases in which apps that depend on the package can break.

Here is a good reference that occurred very recently.
http://www.haneycodes.net/npm-left-pad-have-we-forgotten-how-to-program/

I think it's great that you like to do things your way in your app - but to encourage the rest of the community to be dependent upon a package that essentially clones what's already available in Laravel is, in my opinion, not the best for the developer community.

Again, what you do with your own code is all you, but I would strongly oppose this package for the rest of the community.

That being said, you can define vanilla functions (although I would not, nor would any other experienced developer familiar with PHP/Laravel, recommend doing so) like this:

``` php
<?php // Code in app/Http/Controllers/MyController.php

namespace App\Http\Controllers;

function vanillaThing($string)
{
    echo $string;
}

class MyController extends Controller
{
    public function __construct()
    {
        vanillaThing('Hello');
    }
}
```

If you are looking for that vanilla function to be available globally, well, you're essentially contradicting the OOP and namespaced-functionality that Laravel was meant to emphasize and leverage.
